Inspection tool to check what numba supports

sklam opened the pull request:
<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

Run with `python -m numba.help.inspector <module/package>`.

For example:

```bash
python -m numba.help.inspector numpy
```

will write to `inspector_output.html`.

The listing section in the docs will show the output from `numba.help.inspector` for `builtins`, `math`, `cmath`, `numpy` modules.  The search box in the doc site will be able to locate the listing given a function name. (try searching `dot`)

sklam commented:
Is it better to put it under `numba.help` instead of `numba.inspect`?

esc commented:
I think `numba.inspect.inspector` sounds convoluted due to the double use of the `inspec` prefix. `numba.help.inspector` sounds somewhat better. 🕵

sklam commented:
Update:

* renamed `numba.inspect` to `numba.help`
* added tests
* add auto-generated docs with the new inspector tool
* PR description updated to reflect change.

stuartarchibald commented on docs/source/conf.py:
```diff
@@ -295,13 +295,25 @@
     gen_lower_listing(os.path.join(basedir,
                                    'developer/autogen_lower_listing.rst'))
 
+    # Run inspector on numpy
```
```suggestion
    # Run inspector on supported packages
```

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,388 @@
+"""
+This file contains `__main__` so that it can be run as a commandline tool.
+
+This file contains function to inspect Numba's support for a given Python
```
```suggestion
This file contains functions to inspect Numba's support for a given Python
```

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,388 @@
+    def write_supported_item(self, modname, itemname, typename, explained, sources):
+        self.print('.. function:: {}.{}'.format(modname, itemname))
+        self.print()
+        # if explained:
```
dead code?

sklam replied:
removed

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,388 @@
+from numba.tests.support import captured_stdout
+
+
+commit = get_versions()['full'].split('.')[0]
```
This unfortunately seems to not work quite right in the docs subdir. From the `numba` top level dir:
```
$ pwd
<path>/numba

$ python -c 'from numba._version import get_versions; print(get_versions())'
{'version': '0.46.0dev0+21.g68c4634', 'full': '68c4634857cd1c97dca3f98ffb22487009546778'}
```
from the `numba/docs` subdir:
```
$ pwd
<path>/numba/docs

$ python -c 'from numba._version import get_versions; print(get_versions())'
{'version': '0.46.0.dev0+21.g68c4634-py3.7-linux-x86_64.egg', 'full': ''}
```
(note key `full` has no value) as a result, at least locally, I cannot get the links to render correctly with the `{commit}` present.

sklam replied:
how odd.  it works for me either way.  i think there'll need a default to fallback to the master branch.

sklam replied:
Fixed to fallback to 'master'

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,388 @@
+    # Try getting the function type
+    source_infos = {}
+    try:
+        nbty = tyct.resolve_value_type(function)
```
Not quite sure where to put this comment but I think it's probably to do with this line... It seems like resolving by module doesn't work particularly well due to aliasing. For example...
```
numpy.ma.core.array()
```
is the NumPy masked array impl which is IIRC not supported at all, and yet this code produces:
```
numpy.ma.core.array()
        defined by <class 'numba.typing.templates.Registry.register_global.<locals>.decorate.<locals>.Template'>
```
and I think this is because it considers functions per module opposed to being API aware. Numba, I think, has all it's functions defined with respect to the package root, so an implementation for `array` can only ever be `np.array`? I also think that this is by convention opposed to enforcement, because one could `from numpy import array` and then write something that matches `array` but its not clear where the `array` came from unless the function is checked against the root. FWIW https://github.com/numba/numba/pull/3942 also wrestles with the same problem .

stuartarchibald replied:
There's also the issue of modules importing functionality from elsewhere and there being repeat definitions as a result, e.g. `np.linalg.linalg.all == np.all`. I'm wondering if this tool should walk the top level module and push discovered functions into a `set`, then subsequent submodules can check to see if a function on their module is already defined and explained elsewhere. This approach is imperfect but may reduce repetition.

Also, perhaps `.tests.*` submodules can be skipped?

sklam replied:
Changes:

 - `tests` modules are skipped
- fixed a bug that causes top-level package API to not show.
- fixed a bug that causes `numpy.ma.core.array` to be marked as supported (the wrong name was used).
- aliases symbols are now indicated with "alias to: <first occurence>".

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,388 @@
+""".strip()
+
+
+def main():
```
Is it worth writing tests for the CLI entry point to make sure args are parsed/rejected etc.?

sklam replied:
Added test for the CLI in https://github.com/numba/numba/pull/4353/commits/49b102159e47f0865f5a8790d341987b664cadf0 andhttps://github.com/numba/numba/pull/4353/commits/7dd346572cdccfdd2d1b0fae04241e252bb39e74

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,388 @@
+    )
+    parser.add_argument(
+        '--file', dest='file', default='inspector_output',
+        help='Output filename. Defaults to "inspector_output"',
```
```suggestion
        help='Output filename. Defaults to "inspector_output.<format>"',
```

stuartarchibald replied:
Might be worth mentioning that this overwrites without warning?

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,388 @@
+
+    def write_unsupported_item(self, modname, itemname):
+        pass
+        # self.print('.. function:: {}.{}'.format(modname, itemname))
```
dead?

sklam replied:
removed

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,388 @@
+
+    def write_statistic(self, stat):
+        if stat.supported == 0:
+            self.print("this module is not supported")
```
```suggestion
            self.print("This module is not supported.")
```

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,429 @@
+        self.print()
+
+        if alias:
+            self.print("   alias to: ``{}``".format(alias))
```
I'm wondering if this ought to be declared first and lined up with the indented block. Locally it seems to render as e.g.:
```
numpy.max()

    * defined by <class 'numba.typing.npydecl.Numpy_redirect_max'>

alias to: numpy.amax
```
in firefox and chromimum (the indent is entirely ignored).

I think this might be a stylistic improvement:
```
numpy.max()

    Alias to: numpy.amax

    * defined by <class 'numba.typing.npydecl.Numpy_redirect_max'>
```

sklam replied:
The indent is only saying that "Alias..." is under the `numpy.max()` item.  But, I can't control it's indent further.  I moved the alias up to be just under the `numpy.max()` line.

stuartarchibald commented on numba/help/inspector.py:
```diff
@@ -0,0 +1,429 @@
+            fmtr = ReSTFormatter(fileobj=fout)
+            _format_module_infos(fmtr, package_name, mods)
+    else:
+        raise ValueError("{} is not supported".format(output_format))
```
```suggestion
        raise ValueError("Output format '{}' is not supported".format(output_format))
```
(with linewrap as the above is over 80chars)

stuartarchibald commented:
Thanks for the fixes @sklam, there's a couple of minor comments to resolve and then this is good to merge.

I also noticed that running this tool on `scipy` produces a convenient list of functions which are just NumPy aliases, an adaptation of this informatin may be useful for `numba-scipy`.

stuartarchibald commented:
@sklam this needs conflicts resolved.

stuartarchibald commented on docs/source/conf.py:
```diff
@@ -295,13 +295,25 @@
+        write_listings(
+            package_name=package,
+            filename=os.path.join(
+                basedir, 'developer/autogen_{}_listing'.format(package),
```
```suggestion
                basedir, 'developer', 'autogen_{}_listing'.format(package),
```
^ should probably be that ?!?

stuartarchibald commented:
CI is now failing for a variety of platforms, failure seems legitimate and related to this PR.
